lib/extras/gain_map.cc

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include <jxl/gain_map.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+
+#include "lib/jxl/base/byte_order.h"
+#include "lib/jxl/color_encoding_internal.h"
+#include "lib/jxl/dec_bit_reader.h"
+#include "lib/jxl/enc_bit_writer.h"
+#include "lib/jxl/fields.h"
+#include "lib/jxl/memory_manager_internal.h"
+
+namespace {
+
+template <size_t N>
+void* FixedSizeMemoryManagerAlloc(void* opaque, size_t capacity);
+void FixedSizeMemoryManagerFree(void* opaque, void* pointer) {}
+
+template <size_t N>
+class FixedSizeMemoryManager {
+ public:
+  FixedSizeMemoryManager() = default;
+
+  JxlMemoryManager* memory_manager() { return &manager_; }
+
+  friend void* FixedSizeMemoryManagerAlloc<N>(void* opaque, size_t capacity);
+
+ private:
+  uint8_t memory_[N + jxl::memory_manager_internal::kAlias];
+  JxlMemoryManager manager_ = {
+      /*opaque=*/this,
+      /*alloc=*/&FixedSizeMemoryManagerAlloc<N>,
+      /*free=*/&FixedSizeMemoryManagerFree,
+  };
+};
+
+template <size_t N>
+void* FixedSizeMemoryManagerAlloc(void* opaque, size_t capacity) {
+  auto manager = static_cast<FixedSizeMemoryManager<N>*>(opaque);
+  if (capacity > N + jxl::memory_manager_internal::kAlias) {
+    return nullptr;
+  }
+  return manager->memory_;
+}
+
+}  // namespace
+
+JXL_BOOL JxlGainMapGetBundleSize(const JxlGainMapBundle* map_bundle,
+                                 size_t* bundle_size) {
+  if (map_bundle == nullptr) return 0;
+
+  FixedSizeMemoryManager<sizeof(jxl::ColorEncoding)> memory_manager;
+  jxl::ColorEncoding internal_color_encoding;
+  jxl::BitWriter color_encoding_writer(memory_manager.memory_manager());
+  if (map_bundle->has_color_encoding) {
+    JXL_RETURN_IF_ERROR(
+        internal_color_encoding.FromExternal(map_bundle->color_encoding));
+    if (!jxl::Bundle::Write(internal_color_encoding, &color_encoding_writer,
```

Comment:
IIRC you mentioned that the function to get the size of a Bundle was broken?
If so, we should probably clean it up somehow.
If not, we should use it here ;-)

Reply:
I think we only have the `MaxBits` function, which I think works, but I'm not sure that in this case will returns the actual number of Bits needed or an non-sharp upper bound. Should we use it nonetheless and then perhaps change the documentation and perhaps even the name to `JxlGainMapMaxBundleSize`?

Reply:
Very intuitively, the visitor to get the size appears to be called "CanEncode"...

Reply:
Now using this. Much better don't need the memory manager anymore in this function now..